gbmi/exp_max_of_n/train.py

```diff
@@ -0,0 +1,223 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import Optional, cast
+
+import numpy as np
+import torch
+from jaxtyping import Float, Integer
+from torch import Tensor
+from torch.utils.data import Dataset, TensorDataset, DataLoader, IterableDataset
+from transformer_lens import HookedTransformer, HookedTransformerConfig
+
+from gbmi.model import (
+    TrainingWrapper,
+    Config,
+    ExperimentConfig,
+    train_or_load_model,
+    DataModule,
+)
+from gbmi.utils import generate_all_sequences, shuffle_data, default_device
+
+
+@dataclass
+class MaxOfN(ExperimentConfig):
+    # Max of n (iterable dataset)
+    n_train_samples: Optional[int] = None  # if none, infinite dataset
+    n_test_samples: int = 1024
+
+    # Max of 2 (full dataset)
+    force_adjacent: bool = (
+        False  # whether to force all adjacent-pair inputs to be in training set
+    )
+    training_ratio: float = 0.7  # fraction of dataset to use for training
+
+    def get_training_wrapper(self):
+        return MaxOfNTrainingWrapper
+
+    def get_datamodule(self):
+        return MaxOfNDataModule
+
+    def get_summary_slug(self, config: Config[MaxOfN]) -> str:
+        return f"MaxOf{config.n_ctx}-{config.train_for[0]}-{config.train_for[1]}{'-adj' if self.force_adjacent else ''}"
+
+
+MAX_OF_2_CONFIG = Config(
+    experiment=MaxOfN(
+        force_adjacent=True,
+    ),
+    n_ctx=2,
+)
+MAX_OF_10_CONFIG = Config(
+    experiment=MaxOfN(
+        n_train_samples=None,
+        n_test_samples=1024,
+    ),
+    n_ctx=10,
+)
+
+
+class MaxOfNTrainingWrapper(TrainingWrapper[MaxOfN]):
+    def __init__(self, config: Config[MaxOfN], model: HookedTransformer):
+        super().__init__(config, model)
+        self.model = model
+        self.config = config
+
+    @staticmethod
+    def build_model(config: Config[MaxOfN]) -> HookedTransformer:
+        simpler_cfg = HookedTransformerConfig(
+            d_model=config.d_model,
+            n_layers=config.n_layers,
+            n_heads=config.n_heads,
+            d_head=config.d_head,
+            n_ctx=config.n_ctx,
+            d_vocab=config.d_vocab,
+            seed=config.seed,
+            attn_only=True,
+            normalization_type=None,
+            # device=default_device(deterministic=config.deterministic),
+        )
+        model = HookedTransformer(simpler_cfg)
+        if config.zero_biases:
+            for name, param in model.named_parameters():
+                if "b_" in name:
+                    param.requires_grad = False
+        return model
+
+    @staticmethod
+    def loss_fn(
+        logits: Float[Tensor, "batch pos d_vocab"],
+        tokens: Integer[Tensor, "batch pos"],
+    ) -> Float[Tensor, ""]:
+        logits = logits[:, -1, :]
+        true_maximum = torch.max(tokens, dim=1)[0]
+        log_probs = logits.log_softmax(-1)
```

Comment:
Do we want to do the more accurate (currently untested)
```suggestion
        # log_softmax is only around to around 2e-7, cf https://github.com/pytorch/pytorch/issues/113708
        # we can get better precision by using log1p
        logits_max_idxs = logits.argmax(dim=-1, keepdim=True)
        logits_centered = logits - logits.gather(dim=-1, index=logits_max_idxs)
        logits_exp = logits_centered.exp()
        # logits_exp[max] will be 1, so we can zero it and use log1p(x) = log(1 + x)
        logits_exp.scatter_(dim=-1, index=logits_max_idxs, 0)
        log_probs = logits_centered - logits_exp.sum(dim=-1, keepdim=True).log1p()
```